### PR TITLE
suggest constraints to specify for export based on generated shape guards

### PR DESCRIPTION
**Pull request by avikchaudhuri**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98463

The design of export API expects constraints to be specified on dynamic dimensions, while assuming all other dimensions are static by default. However a user who wishes to export a model may not be fully familiar with the code to plan what to specify.

This diff provides support for discovering constraints to specify. The basic idea is to take the set of generated shape guards and convert them into appropriate constraints. However, we usually generate a LOT of shape guards, and there is often a LOT of redundancy in them. Thus, we also need to simplify the guards so that our suggested constraints are concise yet capture the information content in the guards.

The algorithm for simplification uses `sympy` under the hood, but very surgically to avoid any risk of blowing up. See comments inline for a full description. Briefly,
1. We consider only univariate inequalities, and among them, solve for equalities first.
2. We substitute these exact solutions to convert multivariate inequalities progressively into univariate.
3. Remaining univariate inequalities are solved using `sympy.solvers.inequalities.reduce_inequalities`.
4. As pre-processing, we also eliminate all `//` and `%` operations to generate a set of linear congruence guards, and solve these using `sympy.ntheory.modular.solve_congruence`.

The results are quite dramatic. For example, an internal model produced several hundreds of guards with `dynamic_shapes=True`, which were pretty much inscrutable for humans. The summary contains around 30 dimensions that were specialized and 3 constraints on dynamic dimensions. The output format looks like this:
```
The following dimensions have been specialized and CANNOT be dynamic.
NOTE: Specializations will happen by default with `assume_static_by_default=True`.
	L['foo']['bar'].size()[0] == 4
        ...
	L['baz']['qux'].size()[3] == 96

The following dimensions CAN be dynamic.
You can use the following code to specify the constraints they must satisfy:
constraints=[
	dynamic_dim(L['blah']['bleh'], 1) == dynamic_dim(L['blah']['bloh'], 1),
        ...,
	2 <= dynamic_dim(L['blah']['bloh'], 1),
]
```

Differential Revision: [D44731747](https://our.internmc.facebook.com/intern/diff/D44731747/)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -2231,14 +2523,16 @@ def _add_guard(self, expr: "sympy.Expr") -> None:
...
-                frame_summaries = TracingContext.get().frame_summary_stack
-                # frame_summaries describes a stack of functions
-                # TODO(avik): It would be better to describe a stack of function calls instead
-                current_loc = TracingContext.get().loc_in_frame
-                # current_loc describes a line in the current frame
-                user_stack = ''.join(traceback.format_list([*frame_summaries, current_loc]))
-                expr = LoggingShapeGuardPrinter(self.var_to_sources).doprint(expr)
-                log.warning(f"Adding shape guard {expr} at \n{user_stack}")
+                tc = TracingContext.get()
+                if tc is not None:
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:2527]: Wait why this change? TC should never be None here.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1905,14 +2190,21 @@ def hint():
...
                 if r.upper != sympy.oo and r.upper < sys.maxsize - 1:
+                    if any(is_dim(source) for source in sources):
+                        dim_constraints.add(sympy.Le(symbol, r.upper))
                     bounds.append(str(r.upper))
                 if len(bounds) > 1:
                     exprs.append(" <= ".join(bounds))
 
+        if torch._dynamo.config.dynamic_shapes and torch._dynamo.config.summarize_dim_constraints:
+            dim_constraints.solve()
+            log.warning(f"Summary of dimension constraints:{dim_constraints.prettify_results()}")
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:2201]: No dynamo config notions in this file please. We already have dynamo source notions here in some places and it kills me.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
         super().__init__(var_to_sources, lambda n: n.name(), var_to_sources)
 
 
+class DynamicDimConstraintPrinter(StrPrinter):
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1346]: Please document this. 

What Is the goal of this printer? Produce guard code? debug source? etc.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+        # solve linear congruences
+        # NOTE(avik): We do not need to solve them for symbols that have already been specialized.
+        reduced_congruences = self.reduce_congruences()
+        for s, congruences in reduced_congruences.items():
+            for congruence in congruences:
+                # any congruence that cannot be checked becomes a dynamic constraint as well
+                if not (s in self._substitutions) or not sympy.checksol(congruence, {s: self._substitutions[s]}):
+                    self._dynamic_results.add(self._dcp.doprint(sympy.Eq(congruence, 0)))
+
+    def prettify_univariate_inequalities(self):
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1577]: not called?

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+            buf += f"\n\t],"
+        return buf + "\n}"
+
+    def prettify_multivariate_inequalities(self):
+        buf = "["
+        for expr in self._multivariate_inequalities:
+            buf += f"\n\t{expr},"
+        return buf + "\n]"
+
+    def prettify_congruences(self):
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1592]: not called?

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+    def prettify_univariate_inequalities(self):
+        buf = "{"
+        for free_symbol, exprs in self._univariate_inequalities.items():
+            buf += f"\n\t{free_symbol}: ["
+            for expr in exprs:
+                buf += f"\n\t\t{expr},"
+            buf += f"\n\t],"
+        return buf + "\n}"
+
+    def prettify_multivariate_inequalities(self):
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1586]: not called?

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -1811,6 +2088,8 @@ def hint():
         #    if we have an input (2, 3), we must show s0*2 == 2 and s1 == 3.
         #    This does a lot of work: it covers duck sizing and equality guards.
         exprs = []
+        dim_constraints = DimConstraints(symbol_to_source, self.var_to_val)
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:2091]: This work is unused if summarize_dim_constraints is False, right?

File: torch/_dynamo/config.py [hunk 8]
```diff
@@ -118,6 +118,10 @@
 # Show a warning for every specialization
 print_specializations = False
 
+# Simplify guards, summarizing static and dynamic constraints on dimensions.
+# NOTE: This only has an effect when dynamic_shapes=True.
+summarize_dim_constraints = False
```
Comment (voznesenskym) [inline torch/_dynamo/config.py:123]: Why is this gated, btw? If it's generally good and able, we should include it unequivocally, right? Or what is the tradeoff space for enabling this?

File: torch/fx/experimental/symbolic_shapes.py [hunk 9]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+        # A solution of a free variable with equalities becomes a substitution.
+        # We use these substitutions to simplify other constraints.
+        self._substitutions = {}
+
+        # In general, constraints may have // and % operations.
+        # Of course, // can be expressed in terms of / and %.
+        # Our inequality solver can handle / but not %. So we need to transform them away.
+        # We do so by using the values of variables as hints to evaluate %.
+        # For soundness we record additional congruence guards and solve them separately.
+        self._var_to_val = var_to_val
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1387]: copy to not accidentally mess with this

File: torch/fx/experimental/symbolic_shapes.py [hunk 10]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+        # We park external equalities between free variables here.
+        self._symbolic_equivalences = []
+
+        # Solutions come in two forms:
+        # - (static) specializations
+        # - (dynamic) inequalities / congruences
+        self._static_results = set()
+        self._dynamic_results = set()
+
+        self._dcp = DynamicDimConstraintPrinter(symbol_to_source)
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1402]: copy symbol_to_source to not accidentally mess with this

File: torch/fx/experimental/symbolic_shapes.py [hunk 11]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+                    self._symbols_with_equalities.add(s)
+                self._univariate_inequalities[s].add(expr)
+
+    def add_equality(self, source, expr):
+        if expr.free_symbols:
+            # these will resolve to either specializations or dynamic equality constraints
+            self._symbolic_equivalences.append((source, expr))
+        else:
+            # specialization, right here
+            self._static_results.add(f"{source.name()} == {expr}")
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1493]: Wait, why only here? It seems like every expr should be checked for this, right?

File: torch/fx/experimental/symbolic_shapes.py [hunk 12]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+            )
+
+        return reduced_congruences
+
+    def solve(self):
+        # as long as there are symbols with equalities, solve for them
+        while(self._symbols_with_equalities):
+            s = self._symbols_with_equalities.pop()
+            exprs = self._univariate_inequalities.pop(s)
+            solution = sympy.solvers.inequalities.reduce_inequalities(exprs, s)
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1537]: Are these 2 structures enforced to be the same size somewhere?

File: torch/fx/experimental/symbolic_shapes.py [hunk 13]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+        # We do not try to (directly) solve inequalities with > 1 free variables.
+        self._multivariate_inequalities = set()
+
+        # We park external equalities between free variables here.
+        self._symbolic_equivalences = []
+
+        # Solutions come in two forms:
+        # - (static) specializations
+        # - (dynamic) inequalities / congruences
+        self._static_results = set()
```
Comment (voznesenskym) [inline torch/fx/experimental/symbolic_shapes.py:1399]: what are the types here? I see you throwing strings in?

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Tests, way more docs please, have a few sus spots to check out, otherwise seems like a great direction

File: torch/_dynamo/config.py (hunk 8 above)
Comment (avikchaudhuri) [reply torch/_dynamo/config.py:123]: I'd like to leave this gated for export only, at least until we have been using it for a while. This is mainly because I have only tested this in the wild on export use cases, and I don't know (a) what performance bar we need to meet for eager (b) what corner cases we could be missing that could blow up these solvers, which we haven't used before.

File: torch/fx/experimental/symbolic_shapes.py (hunk 10 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1414]: We're not messing with it.

File: torch/fx/experimental/symbolic_shapes.py [hunk 14]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+
+            # NOTE(avik): This is exactly equivalent to rewriting b // d as (b - (b % d)) / d
+            # and eliminating b % d as above.
+            base, divisor = args
+            base, divisor = self.rewrite_with_congruences(s, base), self.rewrite_with_congruences(s, divisor)
+            mod_reduced = base.subs(self._var_to_val) % divisor.subs(self._var_to_val)
+            congruence = (base - mod_reduced) % divisor
+            if congruence != 0:
+                self._congruences[s].add(congruence)
+            return (base - mod_reduced) / divisor
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1467]: Trust upstream.

File: torch/fx/experimental/symbolic_shapes.py (hunk 11 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1505]: Could you elaborate? We have an assertion in `add()`, we don't expect non-true non-false exprs with no free symbols.

File: torch/fx/experimental/symbolic_shapes.py [hunk 15]
```diff
@@ -1341,6 +1343,279 @@ def __init__(self, var_to_sources):
...
+            substitution = {s: modulus * sympy.Symbol("tmp", integer=True) + remainder}
+            reduced_congruences[s].update(
+                congruence for congruence in congruences_to_check
+                if not sympy.checksol(congruence, substitution)
+            )
+
+        return reduced_congruences
+
+    def solve(self):
+        # as long as there are symbols with equalities, solve for them
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1549]: No, added comment

File: torch/fx/experimental/symbolic_shapes.py (hunk 12 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1554]: Yes, added comment.

File: torch/fx/experimental/symbolic_shapes.py [hunk 16]
```diff
@@ -1718,21 +1993,23 @@ def produce_guards(
...
         symbol_to_constraints = collections.defaultdict(list)
-        dynamic_sources = []
         constraint_violations = []
 
         def record_constraint_violation(msg_fn):
             assert callable(msg_fn)
             constraint_violations.append(msg_fn)
 
+        def is_dim(src):
+            return isinstance(src, TensorPropertySource) and src.prop is TensorProperty.SIZE
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:2006]: Elsewhere in the same file we specifically have this pattern (e.g., see `track_symint`). I don't see an abstraction here that's leaking.

File: torch/fx/experimental/symbolic_shapes.py (hunk 7 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:2086]: Yeah. I thought about hiding it but it's not anything heavy that we don't do already (e.g., we're not calling any solvers when initializing).

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:2196]: How would you suggest I do this, then, assuming I do want to gate it with a flag?

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py]: Cut. I remember running into a problem here, maybe in some other context...

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py]: Cut

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py]: Cut

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py]: Cut

File: torch/fx/experimental/symbolic_shapes.py (hunk 9 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py]: We're evidently not messing with it. Read-only for substitutions.

File: torch/fx/experimental/symbolic_shapes.py (hunk 13 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py]: Added types in init.

**Comment by avikchaudhuri** [comment]
@voznesenskym PTAL, added more docs and tests.

File: torch/fx/experimental/symbolic_shapes.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2006]: I'm not sure why SIZE should be treated specially here but not the other sizevars.

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2196]: Well, the other way we've been doing it is passing these in inside `__init__`. BTW there's no point in testing `dynamic_shapes`, you'll never hit ShapeEnv without it True.

File: torch/fx/experimental/symbolic_shapes.py [hunk 17]
```diff
@@ -1820,6 +2094,8 @@ def hint():
                     source == symbol_to_source[expr][0]
                 ):
                     continue
+                if is_dim(source):
+                    dim_constraints.add_equality(source, expr)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2098]: It's kind of annoying that you have to remember to do a dim_constraints.add anywhere you push to exprs but I can't think way of a good way to force someone to do this

File: torch/fx/experimental/symbolic_shapes.py [hunk 18]
```diff
@@ -1341,6 +1343,274 @@ def __init__(self, var_to_sources):
...
+    - Instead of Eq(_, _), Mod(_, _), etc. it prints _ == _, _ % _, etc.
+
+    We use this to suggest code for specifying dynamic dim constraints.
+    """
+    def __init__(self, symbol_to_source):
+        super().__init__()
+        self.symbol_to_source = symbol_to_source
+
+    def print_source(self, source) -> str:
+        return f"dynamic_dim({source.base.name()}, {source.idx})"
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1359]: This is going to print L['x'] instead of x, I think

File: torch/fx/experimental/symbolic_shapes.py (hunk 17 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2098]: Actually, can't we just get rid of expr entirely, and use dim_constraints to generate the guards in the end? Why can't we do this?

File: torch/fx/experimental/symbolic_shapes.py [hunk 19]
```diff
@@ -1341,6 +1343,274 @@ def __init__(self, var_to_sources):
...
+            self._dynamic_results.add(f"{self._dcp.print_source(source)} == {self._dcp.doprint(expr)}")
+
+        # solve linear congruences
+        # NOTE(avik): We do not need to solve them for symbols that have already been specialized.
+        reduced_congruences = self.reduce_congruences()
+        for s, congruences in reduced_congruences.items():
+            for congruence in congruences:
+                # any congruence that cannot be checked becomes a dynamic constraint as well
+                if not (s in self._substitutions) or not sympy.checksol(congruence, {s: self._substitutions[s]}):
+                    self._dynamic_results.add(self._dcp.doprint(sympy.Eq(congruence, 0)))
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1594]: Are you at all willing to write a translation validator for your solving here via Z3? Based on the discussion at https://github.com/pytorch/pytorch/pull/96616#issuecomment-1484839906

This would give a lot more confidence than just the tests.

File: test/test_dynamic_shapes.py [hunk 20]
```diff
@@ -816,5 +833,944 @@ def is_complex(x):
...
+            Ne(3 * s / 2, 3),
+        }
+        solution = reduce_inequalities(exprs, s).as_set()
+        self.assertEqual(solution, Interval.Ropen(4, 16))
+
+        exprs.add(Eq(s / 2, 4))
+        solution = reduce_inequalities(exprs, s).as_set()
+        self.assertEqual(solution, {8})
+
+    def test_dim_constraints_solve_full(self):
```
Comment (ezyang) [inline test/test_dynamic_shapes.py:878]: These constraints were lifted from somewhere, right? Seems worth mentioning where you got it from.

**Review by ezyang — APPROVED** [review]
I have not closely reviewed the simplification algorithm.

But I guess I don't see any reason not to merge this. The current proposed use case doesn't even need to be sound, since you'll just blast in the constraints and then we will resolve the export according to the new constraints.

File: torch/fx/experimental/symbolic_shapes.py (hunk 19 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1594]: I'm up for it. The best I do here is ensure that at least the concrete values are solutions of original and simplified. But yeah, no equivalence / implication checks.

File: torch/fx/experimental/symbolic_shapes.py (hunk 17 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:2098]: I don't see any reason not to do it, other than wanting to limit the blast radius in one go. There's some parallel simplification that we could get rid of.

File: torch/fx/experimental/symbolic_shapes.py (hunk 18 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1359]: With export, you kind of want to do a rewrite, but not in the way you expect. Suppose you have

```
def f(x, y):
  ...

inp = [torch.randn(3), torch.randn(3)]
torch._dynamo.export(f, inp)
```

for the result to be directly actionable, you have to directly say `dynamic_dim(inp[0], 0)` e.g.

Your current message falls in an uncanny valley, where it sort of looks like it is copy pastable, but it's not. It's better to go all the way, or not at all.

File: torch/fx/experimental/symbolic_shapes.py (hunk 17 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2098]: OK, happy to agree this as PoR and then ship PR as is. But I don't want to turn it on "for real" without translation validation.

**Review by voznesenskym — APPROVED** [review]
Thanks for writing this.

Consider taking the code from https://github.com/pytorch/pytorch/pull/94694 and writing a z3 verifier

**Review by voznesenskym — APPROVED** [review]
Thanks for writing this.

Consider taking the code from https://github.com/pytorch/pytorch/pull/94694 and writing a z3 verifier
